### PR TITLE
DBZ-3577: Use executed gtid as incremental snapshot watermarks

### PR DESCRIPTION
**Pull request by kgalieva**
The read-only incremental snapshot implementation for MySQL using executed GTID set as low/high watermarks

### CODE REVIEW

**Comment by gunnarmorling** [comment]
Thanks a lot, @kgalieva!

@jpechane, could you take a first look?

**Comment by jpechane** [comment]
@kgalieva Hi, thanks a lot for the PR. I'd keep the last word to @gunnarmorling but I personally would prefer slightly different code architecture.
I'd like to see the `MySqlIncrementalSnapshotChangeEventSource` be renamed to `ReadOnlyMySqlIncrementalSnapshotChangeEventSource`. It will not inherit from `SignalBasedIncrementalSnapshotChangeEventSource` but there will be a common abstract parent class that will define the shared logic and `SignalBasedIncrementalSnapshotChangeEventSource` will contain the signal based logic and the new MySQL would contain the GTID based logic.

**Comment by gunnarmorling** [comment]
Hey @jpechane, @kgalieva, where do we stand with this one? Is there a way for testing this new implementation?

**Comment by kgalieva** [comment]
I'm going to test it on my side, I'll post an update as soon as it's done

File: debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+import io.debezium.util.Threads;
+import io.debezium.util.Threads.Timer;
+
+@NotThreadSafe
+public abstract class AbstractIncrementalSnapshotChangeEventSource<T extends DataCollectionId> implements IncrementalSnapshotChangeEventSource<T> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractIncrementalSnapshotChangeEventSource.class);
+
+    private Map<Struct, Object[]> window = new LinkedHashMap<>();
+    private CommonConnectorConfig connectorConfig;
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java:53]: Can it be final?

File: debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+            return;
+        }
+        if (key instanceof Struct) {
+            if (window.remove((Struct) key) != null) {
+                LOGGER.info("Removed '{}' from window", key);
+            }
+        }
+    }
+
+    protected abstract void emitWindowOpen() throws SQLException;
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java:146]: A JavaDoc comment of this method (and the next one) would be nice, given their framework-kind nature.

**Review by gunnarmorling — COMMENTED** [review]
Thanks a lot, @kgalieva! Some minor comments inline.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlOffsetContext.java [hunk 3]
```diff
@@ -209,9 +210,15 @@ public MySqlOffsetContext load(Map<String, ?> offset) {
...
             }
             long binlogPosition = longOffsetValue(offset, SourceInfo.BINLOG_POSITION_OFFSET_KEY);
-
+            IncrementalSnapshotContext<TableId> incrementalSnapshotContext;
+            if (connectorConfig.isReadOnlyConnection()) {
+                incrementalSnapshotContext = ReadOnlyMySqlIncrementalSnapshotContext.load(offset);
+            }
+            else {
+                incrementalSnapshotContext = IncrementalSnapshotContext.load(offset);
+            }
```
[the remark this answers is not in the document]
Comment (kgalieva) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlOffsetContext.java:219]: yes, it is safe, the implementations are interchangeable

**Comment by kgalieva** [comment]
Tested this PR

**Comment by gunnarmorling** [comment]
Thanks a lot, @kgalieva! Hey @jpechane, wanna take another look?

**Comment by jpechane** [comment]
@kgalieva Thanks a lot, it is definitely exactly in the way I was looking for! I'll provide more detail review later. Just top of my head - please convert `IncrementalSnapshotContext` to interface too with two implementations - signal and mysqlreadonly.

**Comment by kgalieva** [comment]
Thanks, @jpechane! I updated this PR and created a follow-up change for the Kafka topic-based signals in case the connection is configured to read-only https://github.com/kgalieva/debezium/pull/1

**Comment by jpechane** [comment]
@kgalieva I think we are almost there! Code is fine from my PoV. Please complete the following clenup tasks before the final merge

- [ ] Please rename `MySqlReadOnlyIncrementalSnapshotChangeEventSource` to `MySqlReadOnlyIncrementalSnapshotChangeEventSource` to highlight it is mysql only
- [ ]  Also please add a comprehensive JavaDoc to this class describing the process how it works.
For signal based approach we have DDD-3 but there is nothing like that fro GTID-based approach.
- [ ] Please extend MySQL documentation with the new config option and also the Incremental Snapshot section
- [ ] Please check the Vitess connector. Is it failing always or only when this chnge was introduced?
- [ ] Create `ReadOnlyIncrementalSnapshotIT` from `IncrementalSnapshotIT` to make the new approach tested as the part of the CI. Note the test can be executed only if run against GTID-enabled database.

Thanks a lot for your effort!

**Comment by jpechane** [comment]
@kgalieva It seems that `io.debezium.connector.mysql.StreamingSourceIT#shouldFailOnSchemaInconsistency` is failing. Could you please take a look at it?

**Comment by jpechane** [comment]
@kgalieva Please check https://github.com/debezium/debezium/pull/2430/checks?check_run_id=3178285429
It seems to me that a single run of `ReadOnlyIncrementalSnapshotIT` takes 7! minutes so the build runtime is almost doubled. Do you know why is that?

**Comment by kgalieva** [comment]
Good catch @jpechane, thank you! The read-only tests were long-running when GTID mode is OFF because each of them was waiting 3 times for 15 seconds to poll the records
https://github.com/debezium/debezium/blob/10ab20a157b4b0ddedfd1306c385be7a64af7d81/debezium-embedded/src/test/java/io/debezium/embedded/AbstractConnectorTest.java#L405
I added a separate test for the expected failure when GTID_MODE is OFF and skipped the rest of the tests.

**Comment by jpechane** [comment]
@kgalieva And it is merged, thanks a lot!

**Pull request merged**